### ISSUE TITLE
Extremely slow interaction, misleading error reports

### ISSUE THREAD

**Issue by EQStack** [brainelectronics/micropython-winbond#11]
While the driver in its current state works, it is extremely slow and also may throw misleading errors:

1) format command returns with an error after 2 seconds
2) writing 10 KiB of data takes about one minute

Maybe the await function is not ideal currently? I found out that sleep(0.1) there leads to very slow speeds and that the timeout of 2 seconds will display busy errors for all longer running tasks.

**Comment by brainelectronics** [comment]
Thank you for reporting this issue! 
Could you give me some more insights to your used code? 
It seems to me that the external flash is not ready or blocked in some other way. That's why the `_await()` function terminated after 2 seconds as it was not able to get a "not busy" state from the flash. It tries to acquire and check this bit with a 100ms delay for 20 times

**Comment by EQStack** [comment]
Thanks for responding! I am using ESP32-C3 in two versions, mini (compatible with Wemos D1) and super mini, both with the latest uPython 1.22.2. There is no other device connected to the MCU. My setup code is pretty straightforward:
```
import machine, winbond
spi = machine.SPI(1, sck=machine.Pin(2), miso=machine.Pin(3), mosi=machine.Pin(4))
w25 = winbond.W25QFlash(spi, cs=machine.Pin(5), software_reset=True)
```
The import initially fails with syntax error in `winbond.py` on line 176, but that is easy to correct. Then everything is initialized and the chip reports itself as type 0x40 with 16 MiB of memory.
However `w25.format()` returns with error after 2 seconds, while the chip needs about half a minute to execute the command (as I tested it), so it is not ready for a long time by desing. Also `os.VfsFat.mkfs(w25)` needs quite a bit more than 10 minutes to finish and writing 15 KiB file takes at least 3 minutes.
The extremely slow writing diappears when `sleep(0.1)` in `_await` is replaced e.g. by `sleep_us(1)` and much higher `trials` count (or there is no busy time limit).  After this change, FAT file system is created in 3 seconds and copying of the mentioned file takes 5 seconds.
However, I do not consider myself proficient enough to be sure this code adjustment is right, and that is why I raised the issue(s).

**Comment by brainelectronics** [comment]
I'be never tested the onboard flash to be honest... I always used an external flash chip connected to the SPI port. 
In case you use the internal flash of the ESP, I'm quite sure the micropython system is the blocking element. 
Could you also post the syntax error of line 176, which I can not reproduce here on my setup. 
Erasing my 16MiB chip takes round about 1 minute, which seems fair for that size. 

May you can also post your complete code here?

**Comment by EQStack** [comment]
I am not sure why you talk about onboard flash. What I am using is an external W25Q chip, 16 MiB of EEPROM, connected to SPI as indicated in the code. The code already posted is complete, what follows is only `format` and `mkfs` commands issued to the external chip.
With the `format` command the driver returns with error after 2 seconds of waiting - but the command is being executed, only it takes about half a minute to complete.
The `mkfs` command takes more than 15 minutes to complete - very slow writing is, I believe, caused by repeated waiting for extremely long 0.1 s in the `_await` loop.
For myself I have disabled throwing the busy error in `_await` loop and changed waiting for only 1 microsecond. The `winbond.py` driver now works flawlessly, with speeds comparable to the onboard EEPROM.

**Comment by brainelectronics** [comment]
I've checked the schematic of the [Waveshare ESP32-C3 Zero](https://www.waveshare.com/wiki/ESP32-C3-Zero)

The Hardware SPI pins are actually not connected to the small pinheader, they should be pins `12, 13, 14` for HSPI or `18, 19, 23` for VSPI of the C3, see [Hardware SPI MicroPython](https://docs.micropython.org/en/latest/esp32/quickref.html#hardware-spi-bus)

May you can try using the [MicroPyhon Software SPI bus](https://docs.micropython.org/en/latest/esp32/quickref.html#software-spi-bus)

```python
spi = machine.SoftSPI(sck=machine.Pin(2), miso=machine.Pin(3), mosi=machine.Pin(4))
w25 = winbond.W25QFlash(spi, cs=machine.Pin(5), software_reset=True)
```

In addition, decreasing the SPI bus speed could also be a solution in case you use jumper cables and a breadboard for your setup. I would start with e.g. `baudrate=100000` (100kHz) an increase it by a factor of 10 for each test run. The default of this lib is `40000000` (40MHz) which is to high if you don't use a custom PCB layout.

**Comment by brainelectronics** [comment]
Hey @EQStack @hstarmans fixed your reported slow writing. It's published as [`0.5.4`](https://github.com/brainelectronics/micropython-winbond/releases/tag/0.5.4)
I'll close this report as completed. Please reopen otherwise
